### DISCUSSION TITLE
sqlite: generate table `WITHOUT ROWID`?

### DISCUSSION THREAD

**Discussion by jakswa** [SeaQL/sea-orm#1672] [Q&A]
Hello, new user, sorry if this is obvious or documented and I missed it.

Sqlite has a `WITHOUT ROWID` feature that is a performance optimization that I hope to use on a small-but-high-throughput table: https://www.sqlite.org/withoutrowid.html

> in some cases, a WITHOUT ROWID table can use about half the amount of disk space and can operate nearly twice as fast. Of course, in a real-world schema, there will typically be secondary indices and/or UNIQUE constraints, and the situation is more complicated. But even then, there can often be space and performance advantages to using WITHOUT ROWID on tables that have non-integer or composite PRIMARY KEYs. 

I am struggling to find support for this in the [`TableCreateStatement` builder](https://docs.rs/sea-orm-migration/0.11.3/sea_orm_migration/prelude/struct.TableCreateStatement.html#method.build). Can I add a custom string on the end of the SQL maybe?

**Comment by billy1624** [comment]
Hey @jakswa, SeaQuery does not support appending additional SQL after the table create statement. But there is a way to get around it. We can build the statement manually and append extra SQL to the very end of the statement.

```rs
#[async_trait::async_trait]
impl MigrationTrait for Migration {
    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
        let create_table = Table::create()
            .table(Cake::Table)
            .col(
                ColumnDef::new(Cake::Id)
                    .integer()
                    .not_null()
                    .auto_increment()
                    .primary_key(),
            )
            .col(ColumnDef::new(Cake::Name).string().not_null())
            .to_owned();

        let conn = manager.get_connection();
        let builder = conn.get_database_backend();
        let mut stmt = builder.build(&create_table);
        write!(stmt.sql, " WITHOUT ROWID");
        conn.execute(stmt).await.map(|_| ())
    }

// ...
```
